### PR TITLE
server,sql/p{arse,lan}: Move handling of KILL statements to plan nodes instead of adhoc regex parsing in server handler.

### PR DESCRIPTION
**Pull request by reltuk**

### CODE REVIEW

File: sql/session.go [hunk 1]
```diff
@@ -724,11 +732,29 @@ func (c *Context) Warn(code int, msg string, args ...interface{}) {
...
 }
 
+// Services are handles to optional or plugin functionality that can be
+// used by the SQL implementation in certain situations. An integrator can set
+// methods on Services for a given *Context and different parts of go-mysql-server
+// will inspect it in order to fulfill their implementations. Currently, the
+// KillConnection service is available. Set these with |WithServices|; the
+// implementation will access them through the corresponding methods on
+// *Context, such as |KillConnection|.
+type Services struct {
```
Comment (reltuk) [inline sql/session.go:754]: I'm going with this for now...we will see how well it works going forward and we can always change it.

The other short-term addition here is something for the `LOAD DATA` plan node to call into mysql.Conn.HandleLoadDataLocalQuery, which transits the remote file from the client to the server. 
https://github.com/dolthub/vitess/blob/main/go/mysql/conn.go#L739. (`HandleLoadDataLocalQuery` will probably change to something like `func (*Conn) LoadFileFromClient(file string) (io.Reader, error)`)

**Comment by reltuk** [comment]
Race detector is amazing and found racey behavior in the new (correct) handling of `KILL CONNECTION` because the query hander could call `rowIter.Close()` before waiting for the goroutine that was calling `rowIter.Next()` to have terminated. Cleaned that up by moving all of the err chan/quit chan stuff to an errgroup.

File: server/handler.go [hunk 2]
```diff
@@ -339,6 +328,9 @@ func (h *Handler) doQuery(
 		}
 	}()
 
+	eg, egCtx := errgroup.WithContext(ctx)
```
Comment (reltuk) [inline server/handler.go:331]: Don't think this change should affect performance much, but will benchmark it in the merge to dolt.

File: server/context.go [hunk 3]
```diff
@@ -40,6 +40,11 @@ func DefaultSessionBuilder(ctx context.Context, c *mysql.Conn, addr string) (sql
 	return sql.NewBaseSessionWithClientServer(addr, client, c.ConnectionID), nil
 }
 
+type Entry struct {
```
Comment (zachmu) [inline server/context.go:43]: Session?

SessionConnection?

File: server/context.go [hunk 4]
```diff
@@ -170,13 +174,25 @@ func (s *SessionManager) NewContextWithQuery(conn *mysql.Conn, query string) (*s
...
 }
 
-// CloseConn closes the connection in the session manager and all its
-// associated contexts, which are cancelled.
+func (s *SessionManager) killConnection(connID uint32) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if entry, ok := s.sessions[connID]; ok {
+		delete(s.sessions, connID)
+		entry.conn.Close()
```
Comment (zachmu) [inline server/context.go:190]: Maybe add a note that we always kill any associated process before calling this (it looks incomplete as is)

**Review by zachmu — APPROVED** [review]
LGTM

**Pull request merged**